libr/bin/p/bin_elf.c

```diff
@@ -492,34 +485,146 @@ static RList* relocs(RBinFile *arch) {
 	 * free detected by ASan that would corrupt the relocation names */
 	r_list_free (imports (arch));
-
-#if 1
-	if ((got_addr = Elf_ (r_bin_elf_get_section_addr) (arch->o->bin_obj, ".got")) == -1) {
-		got_addr = Elf_ (r_bin_elf_get_section_addr) (arch->o->bin_obj, ".got.plt");
-		got_addr = 0;
+	if ((got_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".got")) == -1) {
+		got_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".got.plt");
+		if (got_addr == -1)
+			got_addr = 0;
+	}
+	if (got_addr < 1 && bin->ehdr.e_type == ET_REL) {
+		got_addr = Elf_(r_bin_elf_get_section_addr) (bin, ".got.r2");
+		if (got_addr == -1)
+			got_addr = 0;
 	}
-
 	if (arch->o) {
-		if (!(relocs = Elf_(r_bin_elf_get_relocs) (arch->o->bin_obj)))
+		if (!(relocs = Elf_(r_bin_elf_get_relocs) (bin)))
 			return ret;
 		for (i = 0; !relocs[i].last; i++) {
-			if (!(ptr = reloc_convert (arch->o->bin_obj,
-					&relocs[i], got_addr)))
+			if (!(ptr = reloc_convert (bin, &relocs[i], got_addr)))
 				continue;
 			r_list_append (ret, ptr);
 		}
 		free (relocs);
 	}
-#endif
+	return ret;
+}
+
+
+static void __patch_reloc (RIO *io, RBinElfReloc *rel, ut64 vaddr) {
+	static int times = 0;
+	times++;
+	switch (rel->type) {
+	case R_X86_64_PC32: //R_386_PC32 both have the same value
+		{
+			char s[64];
+			ut64 num = vaddr - (rel->rva + 4);
+			num  = ((num << 8) & 0xFF00FF00 ) | ((num >> 8) & 0xFF00FF);
+			//if s is equal to 0x42d we should get 0x042d that is why %04
+			snprintf (s, sizeof (s), "%04"PFMT64x, num);
+			ut8 *buf = malloc (strlen (s) + 1);
+			int len = r_hex_str2bin (s, buf);
+			r_io_write_at (io, rel->rva, buf, len);
+			free (buf);
+		}
+		break;
+	case R_X86_64_32S:
+		{
+			char s[64];
+			st32 num = r_swap_st32(vaddr);
+			snprintf (s, sizeof (s), "%08x", num);
+			ut8 *buf = malloc (strlen (s) + 1);
+			int len = r_hex_str2bin (s, buf);
+			r_io_write_at (io, rel->rva, buf, len);
+			free (buf);
+		}
+		break;
+	case R_X86_64_64: //R_386_32
+		{
+			char s[64];
+			ut64 num = r_swap_ut64(vaddr);
+			snprintf (s, sizeof (s), "%08"PFMT64x, num);
+			ut8 *buf = malloc (strlen (s) + 1);
+			int len = r_hex_str2bin (s, buf);
+			r_io_write_at (io, rel->rva, buf, len);
+			free (buf);
+		}
+		break;
+	default:
+		eprintf ("No relocation for %d\n", rel->type);
+		break;
+	}
+}
+
+static RList* patch_relocs(RIO *io, RBin *b) {
+	RList *ret = NULL;
+	RBinReloc *ptr = NULL;
+	RBinObject *obj = NULL;
+	struct Elf_(r_bin_elf_obj_t) *bin = NULL;
+	RIOSection *g = NULL, *s = NULL;
+	RListIter *iter;
+	RBinElfReloc *relcs = NULL;
+	int i;
+	ut64 n_off, n_vaddr, vaddr, size, sym_addr = 0, offset = 0;
+	if (!b || !io)
+		return NULL;
+	obj = r_bin_cur_object (b);
+	if (!obj)
+		return NULL;
+	bin = obj->bin_obj;
+	if (bin->ehdr.e_type != ET_REL)
+		return NULL;
+	if (!io->cached) {
+	    	eprintf (Color_YELLOW"[ELF OBJ] run r2 with -e io.cache=true to patch the binary and get better disassm"Color_RESET"\n");
```

Comment:
This line contains color escape codes. so it should be checking for scr.color or print->flags... but i think that its ok if we just left it without Color_* .. also, try to make the string shorter like this:

`Warning: run r2 with -e io-.cache=true to fix relocations in disassembly.`